wrappers/internal/service_status.go

```diff
@@ -68,60 +78,168 @@ func (s *ServiceStatus) IsEnabled() bool {
-			continue
-		}
 		svc, activators := SnapServiceUnits(app)
-		allUnits = append(allUnits, svc)
-		allUnits = append(allUnits, activators...)
+		if app.DaemonScope == snap.SystemDaemon {
+			sys = append(sys, svc)
+			sys = append(sys, activators...)
+		} else if app.DaemonScope == snap.UserDaemon {
+			usr = append(usr, svc)
+			usr = append(usr, activators...)
+		}
 	}
-	return allUnits
+	return sys, usr
 }
 
 func serviceIsSlotActivated(app *snap.AppInfo) bool {
 	return len(app.ActivatesOn) > 0
 }
 
-func QueryServiceStatusMany(sysd systemd.Systemd, apps []*snap.AppInfo) ([]*ServiceStatus, error) {
-	allUnits := appServiceUnitsMany(apps)
-	unitStatuses, err := sysd.Status(allUnits)
+var userSessionQueryServiceStatusMany = func(units []string) (map[int][]client.ServiceUnitStatus, map[int][]client.ServiceFailure, error) {
+	// Avoid any expensive call if there are no user daemons
+	if len(units) == 0 {
+		return nil, nil, nil
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
+	defer cancel()
+	cli := client.New()
+	return cli.ServiceStatus(ctx, units)
+}
+
+func clientUnitStatusToSystemdUnitStatus(unitStatus client.ServiceUnitStatus) *systemd.UnitStatus {
+	return &systemd.UnitStatus{
+		Daemon:           unitStatus.Daemon,
+		Id:               unitStatus.Id,
+		Name:             unitStatus.Name,
+		Names:            unitStatus.Names,
+		Enabled:          unitStatus.Enabled,
+		Active:           unitStatus.Active,
+		Installed:        unitStatus.Installed,
+		NeedDaemonReload: unitStatus.NeedDaemonReload,
+	}
+}
+
+func findMatchingStatus(name string, services []client.ServiceUnitStatus) *systemd.UnitStatus {
+	for _, svc := range services {
+		if svc.Name == name {
+			return clientUnitStatusToSystemdUnitStatus(svc)
+		}
+	}
+	return nil
+}
+
+func queryUserServiceStatusMany(apps []*snap.AppInfo, units []string) (map[int][]*ServiceStatus, error) {
+	usrUnitStss, _, err := userSessionQueryServiceStatusMany(units)
 	if err != nil {
 		return nil, err
 	}
 
-	var appStatuses []*ServiceStatus
-	var statusIndex int
+	constructStatus := func(app *snap.AppInfo, uid int) *ServiceStatus {
+		primarySvcName, activators := SnapServiceUnits(app)
+		primarySvcUnit := findMatchingStatus(primarySvcName, usrUnitStss[uid])
+		if primarySvcUnit == nil {
+			return nil
+		}
+
+		svcSt := &ServiceStatus{
+			name:        app.Name,
+			user:        true,
+			service:     primarySvcUnit,
+			slotEnabled: serviceIsSlotActivated(app),
+		}
+		if len(activators) > 0 {
+			for _, act := range activators {
+				actSvcUnit := findMatchingStatus(act, usrUnitStss[uid])
+				if actSvcUnit == nil {
+					return nil
+				}
+				svcSt.activators = append(svcSt.activators, actSvcUnit)
+			}
+		}
+		return svcSt
+	}
+
+	// For each user we have results from, go through services and build a list of service results
+	svcsStatusMap := make(map[int][]*ServiceStatus)
+	for uid := range usrUnitStss {
+		var svcs []*ServiceStatus
```

Comment:
wouldn't it make sense to create a map name -> systemd.UnitStatus from usrUnitStts[uid] instead of using  findMatchingStatus ?

Reply:
super nice suggestion, changed this